src/Allocator/Allocator.hs

```diff
@@ -0,0 +1,27 @@
+module Allocator.Allocator where
+
+import Allocator.Process
+import Foundation
+import Control.Effect.Error
+import Control.Effect.State
+import Data.HashMap (Map)
+
+-- map of process id to its pages
+type PMap = Map Pid Mem
+
+data AllocError = OutOfMemory deriving (Eq, Ord)
+
+type AllocSig sig m = ( Has (State PMap) sig m
+                      , Has (Throw AllocError) sig m
+                      , Has (Catch AllocError) sig m)
+
+
+class Allocator a where
+    alloc   :: Has (State PMap) sig m => a -> [Word8] -> m (Maybe (Word8))
```

Comment:
`a -> [Word8] -> ...` need to be `a -> Word8`. Also, maybe this parameter a will be for pointers and not for `Process`(see above comment about `Process` instances passing implicitly). Then we can implement `Allocator` for `Pointer64` and `Pointer32` or smt like this. Just an idea, though, need further discussion...